structs/map.go

```diff
@@ -0,0 +1,67 @@
+package structs
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+var (
+	// ErrTagEmpty is returned when an empty tag is passed to Map.
+	ErrTagEmpty = errors.New("tag cannot be empty")
+	// ErrInvalidInputType is returned when Map receives an input that is not a struct or a pointer to a struct.
+	ErrInvalidInputType = errors.New("invalid input type")
+)
+
+// Map converts the given struct to a map[string]any. It uses the given tag to identify the key for each map field.
+// If no tag is present in the Struct field, the field is omitted.
+//
+// Considering the following struct:
+//
+//	type Test struct {
+//		Field1 string `test:"field_1"`
+//		Field2 int `test:"field_2"`
+//		Field3 any `test:"field_3"`
+//	}
+//
+// And the following variable:
+//
+//	input := Test{
+//		Field1: "test",
+//		Field2: 1,
+//		Field3: nil,
+//	}
+//
+// Map(input, "test") will return a map as follows:
+//
+//	map[string]any{
+//		"field_1": "test",
+//		"field_2": 1,
+//		"field_3": nil,
+//	}
+//
+// If Map is called with a non-existent tag, an empty map will be returned.
+//
+//	m := Map(input, "not_found")
+//	len(m) == 0 (true)
+func Map(s any, tag string) (map[string]any, error) {
```

Comment:
I don't have a strong opinion about this, but why not take the `json` approach and have the tag always be `map`? 

Do you see a use case where you want to support multiple tags (so as to convert a single struct to multiple potential maps), or where the tag can change based on the feature?

In addition to simplifying calls to the `Map` function, it also allows for this function to convert all fields in a struct by default, reducing the amount of work necessary to integrate existing structs and more importantly structs you don't have control over.

I suggest that the tag is always set to `map` (or some other value you prefer), and allow for 3 possible values:

* No tag → Use the field name converted to snake case to. 
  * E.g. `Field1 string` would be converted to `field_1`: `<value>`. 
  * You can use an external library to convert the name.
* Tag with `-` as value → Ignore this field when converting to a map.
* Tag with any other value → Use the contents of the tag as the map key. 
  * E.g. ``Field1 string `map:"test"` ``would be converted to `test`: `<value>`.

Reply:
> I don't have a strong opinion about this, but why not take the json approach and have the tag always be map?

My plan was to use the `sendgrid` tag for sendgrid dynamic template structs. I agree that we could go with a static struct, just like https://github.com/fatih/structs does (it uses `structs`). I was just future proofing.

> No tag → Use the field name converted to snake case as the map key.
E.g. Field1 string would be converted to field_1: <value>.
You can use an external library to convert the name.
Tag with - as value → Ignore this field when converting to a map.
Tag with any other value → Use the contents of the tag as the map key.
E.g. Field1 string `map:"test"` would be converted to test: <value>.

These changes will be implemented when switching to https://github.com/fatih/structs